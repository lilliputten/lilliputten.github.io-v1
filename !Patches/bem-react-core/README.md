> @since 2018.02.23, 23:35
> @version 2018.02.23, 23:35

# Pathcing for dev version of react

Modifiyng the `main` entry in `package.json`.

## Changing:

Pathing root: `node_modules`

Package: `bem-react-core@1.0.0-rc.6`

Path `bem-react-core`:
  - `package.json`
