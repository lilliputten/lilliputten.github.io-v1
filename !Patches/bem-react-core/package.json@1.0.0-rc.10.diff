--- package.json.orig	2018-11-12 21:37:31.618494100 +0300
+++ package.json	2018-11-12 21:38:25.225717700 +0300
@@ -88,7 +88,7 @@
       "git add"
     ]
   },
-  "main": "dist/react.min.js",
+  "main": "dist/react.js",
   "name": "bem-react-core",
   "peerDependencies": {
     "preact": "^8.2.6",
