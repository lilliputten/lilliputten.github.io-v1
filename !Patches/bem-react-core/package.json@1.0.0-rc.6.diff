--- package.ORIG.json	Sun Jan 28 23:57:40 2018
+++ package.json	Fri Feb 23 23:28:59 2018
@@ -84,7 +84,7 @@
     "core"
   ],
   "license": "MPL-2.0",
-  "main": "dist/react.min.js",
+  "main": "dist/react.js",
   "name": "bem-react-core",
   "peerDependencies": {
     "preact": "^8.2.6",
