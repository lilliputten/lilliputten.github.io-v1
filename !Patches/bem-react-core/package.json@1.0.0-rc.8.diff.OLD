--- package.json.orig	Tue Mar 13 00:10:18 2018
+++ package.json	Tue Mar 13 01:22:20 2018
@@ -117,7 +117,7 @@
       "git add"
     ]
   },
-  "main": "dist/react.min.js",
+  "main": "dist/react.js",
   "maintainers": [
     {
       "name": "yeti-or",
