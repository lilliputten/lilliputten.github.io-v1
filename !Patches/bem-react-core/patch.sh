#!/sbin/sh
# @since 2018.02.23, 23:35
# @version 2018.02.23, 23:35

ORIGPATH=`pwd`

# Patching root (finding this folder in parents recursively).
while [ ! -d "$ROOT" ]; do
    echo "? $ROOT"
    ROOT="../$ROOT"
done

# Target path relative to root
PATCHCD="bem-react-core"

echo "patching in $ROOT/$PATCHCD"
cd "$ROOT/$PATCHCD"

# Patching file
PATCHFILE="package.json@1.0.0-rc.6.diff"
echo "patching with $PATCHFILE"
cp "$ORIGPATH/$PATCHFILE" .
patch -lb < "$PATCHFILE"

cd $ORIGPATH

