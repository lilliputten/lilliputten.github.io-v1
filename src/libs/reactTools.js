/**
 * @module reactTools
 * @author lilliputten <lilliputten@yandex.ru>
 * @since 2018.03.10, 03:56
 * @version 2018.03.10, 04:44
 *
 * TODO:
 *
 * - 2018.03.10, 19:18 -- Find components with modName/modVal & elems.
 * - 2018.03.10, 23:04 -- Move DOM methods to `react-traverse`?
 *
 */

import React from 'react';
import ReactDom from 'react-dom';
import jquery from 'jquery';

const isArray = Array.isArray

const reactTools = {

  // Traverse:

  /** findParentBlock ** {{{ Find closest parent component
   * @param {ReactObject} comp
   * @param {String} compClassName
   * @return {ReactObject}
   * @see [Element.closest() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)
   * @see [Document.getElementsByClassName() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName)
   * @see [ReactDOM - React](https://reactjs.org/docs/react-dom.html#finddomnode)
   */
  findParentBlock : function (comp, compClassName) {
    const selector = '.' + compClassName
    const foundDom = this.applyQuerySelector(comp, 'closest', selector)
    const foundComp = foundDom && this.getDomComponent(foundDom)
    return foundComp;
  },/*}}}*/
  /** findChildBlock ** {{{ Find closest parent component
   * @param {ReactObject} comp
   * @param {String} compClassName
   * @return {ReactObject}
   * @see [Element.closest() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)
   * @see [Document.getElementsByClassName() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName)
   * @see [ReactDOM - React](https://reactjs.org/docs/react-dom.html#finddomnode)
   */
  findChildBlock : function (comp, compClassName) {
    const selector = '.' + compClassName
    const foundDom = this.applyQuerySelector(comp, 'find', selector)
    const foundComp = foundDom && this.getDomComponent(foundDom)
    return foundComp;
  },/*}}}*/

  // DOM:

  /** applyQuerySelector ** {{{ Apply querySelector lookup method
   * @param {Enzyme:ReactWrapper|ReactComponent|jQueryCollection} target
   * @param {String} op - querySelector operator (eg, `find`, `closest`)
   * @param {String} selector - dom selector (`.className`)
   * @return {jQueryCollection|null)
   */
  applyQuerySelector : function (target, op, selector) {
    let dom = this.getComponentDom(target) // (typeof target.getDOMNode === 'function') ? target.getDOMNode : (target instanceof React.Component) ? ReactDom.findDOMNode(target) : target
    const isJQuery = dom instanceof jquery;
    if (!isJQuery) {
      dom = jquery(dom)
    }
    if (typeof dom[op] === 'function') {
      return dom[op].call(dom, selector)
    }
  },/*}}}*/

  /** getComponentDom ** {{{ Get DOM Node for given component
   * @param {Enzyme:ReactWrapper|ReactComponent|jQueryCollection} target
   * @return {DomObject}
   */
  getComponentDom : function (target) {
    let result = target;
    if (typeof target.getDOMNode === 'function') {
      result = target.getDOMNode();
    }
    else if (target instanceof React.Component) {
      result = ReactDom.findDOMNode(target);
    }
    return result;
  },/*}}}*/
  /** getDomComponent ** {{{ Get react component instance for given DOM node
   * @param {DomObject} dom
   * @return {ReactObject}
   * @see [javascript - React - get React component from a child DOM element? - Stack Overflow](https://stackoverflow.com/questions/24462679/react-get-react-component-from-a-child-dom-element)
   */
  getDomComponent : function (dom) {
    if (dom && dom != null && typeof dom === 'object') {
      // Array|Collection...
      if (isArray(dom) || dom instanceof jquery) {
        dom = dom[0];
      }
      // enzyme ReactWrapper...
      if (typeof dom.instance === 'function') {
        dom = dom.instance();
      }
      if (typeof dom === 'object') {
        for (let key in dom) {
          if (key.startsWith('__reactInternalInstance')) {
            const domItem = dom[key]
            // Old (stackoverflow) code (react15.3.0): `comp = domItem._currentElement._owner._instance`
            // NOTE 2018.03.10, 04:44 -- Works with react16
            const comp = domItem && domItem.return && domItem.return.stateNode
            return comp;
          }
        }
      }
    }
  },/*}}}*/

  // url:

  /** getRelativeUrl ** {{{ Get url relative to current document
   * @param {String} url
   * @return {String}
   */
  getRelativeUrl : function (url) {
    if (typeof window === 'object' && window.location && window.location.origin && url.startsWith(window.location.origin)) {
      const len = window.location.origin.length;
      url = url.substr(len);
    }
    return url;
  },/*}}}*/

  // utils:

  /** delay ** {{{ DEBUG: Timeout
   * TODO: Move to `commonTools`
   * @param {Number} [timeout=1000]
   * @param {*} [data]
   * @return {Promise}
   */
  delay(timeout=1000, data){
    return new Promise(function(resolve, reject){
      setTimeout(function(){
        resolve(data);
      }, timeout)
    });
  },/*}}}*/


};

export default reactTools;

