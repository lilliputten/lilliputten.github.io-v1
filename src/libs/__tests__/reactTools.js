import React, { Fragment } from 'react';
import { decl, Bem } from 'bem-react-core';

import Enzyme, { mount/* , shallow */ } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import reactTools from 'libs/reactTools';

Enzyme.configure({ adapter: new Adapter() })

describe('reactTools', () => {

  // Test components...
  /** Test ** {{{ */
  const Test = decl({
    block : 'Test',
    willInit() {
      this.testMethod = this.testMethod.bind(this);
    },
    testMethod() {
      return 'Test';
    },
    content() {
      return (
        <Fragment>
          <Bem block="Button" tag="button">
            <Bem elem="Text">Go!</Bem>
          </Bem>
        </Fragment>
      );
    },
  }).applyDecls()/*}}}*/
  /** TestWrapper ** {{{ */
  const TestWrapper = decl({
    block : 'TestWrapper',
    willInit() {
      this.testMethod = this.testMethod.bind(this);
    },
    testMethod() {
      return 'TestWrapper';
    },
    content() {
      return (
        <Test/>
      );
    },
  }).applyDecls()/*}}}*/

  /** DOM ** {{{ */
  describe('DOM', () => {

    let testComp;

    beforeAll(() => {
      testComp = mount(
        <TestWrapper />
      );
    })

    afterAll(() => {
      testComp.unmount();
    })

    it('match snapshot', () => {
      expect(testComp).toMatchSnapshot();
    })

    let childWrapper;
    it('find child', () => {
      // get ReactWrapper object by querySelector
      childWrapper = testComp.find('.Test');
      expect(childWrapper).toHaveLength(1);
    })

    it('getComponentDom', () => {
      // get react component from ReactWrapper
      const comp = reactTools.getDomComponent(childWrapper);
      expect(comp).toBeInstanceOf(React.Component);
    })

    let childComp;

    it('findChildBlock', () => {
      // hi-level child block/component finding method
      childComp = reactTools.findChildBlock(testComp, 'Test');
      expect(childComp).toBeInstanceOf(React.Component);
    })

    it('findParentBlock', () => {
      // hi-level parent block/component finding method
      const parentComponent = reactTools.findParentBlock(childComp, 'TestWrapper');
      expect(parentComponent).toBeInstanceOf(React.Component);
    })

  })/*}}}*/

})

