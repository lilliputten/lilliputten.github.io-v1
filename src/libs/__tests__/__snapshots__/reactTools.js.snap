// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactTools DOM match snapshot 1`] = `
<TestWrapper>
  <div
    className="TestWrapper"
    style={Object {}}
  >
    <Test>
      <div
        className="Test"
        style={Object {}}
      >
        <Bem
          block="Button"
          tag="button"
        >
          <button
            className="Button"
          >
            <Bem
              elem="Text"
            >
              <div
                className="Button-Text"
              >
                Go!
              </div>
            </Bem>
          </button>
        </Bem>
      </div>
    </Test>
  </div>
</TestWrapper>
`;
